Make Parser & Tokenizer generic over Dialect, remove `test_utils` module

not-my-profile opened the pull request:
* rustc monomorphizes generics which is more efficient than
  trait object methods which are dispatched dynamically
    
* the Parser struct no longer has a lifetime, making it overall
   more convenient (e.g it now is Send, Sync & UnwindSafe)

For details see the commit messages.

If you like these changes, please merge them without squashing.

alamb commented on tests/test_utils.rs:
```diff
@@ -26,31 +21,79 @@ use alloc::{
+    };
+}
+
+pub trait Parse {
```
I think you are using this trait so you can have a `dyn Parse` and dispatch dynamically at runtime what dialect is being used. This is likely something other users of the crate might want / might be using (dynamically pick a dialect).

What would you think about putting this trait in the main  crate (or perhaps even calling it `Parser` and changing the struct to `ParserImpl`)?

not-my-profile replied:
I highly doubt that any API users actually need `Parser` trait objects. Note that with generics it's still possible to "dynamically" choose which dialect is being used at runtime, you don't need trait objects for that. As an example for that see the updated `examples/cli.rs`.

alamb replied:
> As an example for that see the updated examples/cli.rs.

Indeed -- I think in the case of `examples/cli.rs` the distinction is that the call to `parse<D>` is now also generic (the generics perolate back up)

alamb reviewed (commented):
Thank you @not-my-profile  -- this is a very nice contribution

I had a few questions related to test changes and the use of `PhantomData` but otherwise this looks like a nice change to me, even though it is backwards incompatible.

cc @Dandandan  @houqp @andygrove and @nickolay

alamb commented:
Looks like there are some CI failures as well

not-my-profile commented on src/lib.rs:
```diff
@@ -45,9 +43,3 @@ pub mod dialect;
-// This is required to make utilities accessible by both the crate-internal
-// unit-tests and by the integration tests <https://stackoverflow.com/a/44541071/1026>
-// External users are not supposed to rely on this module.
-pub mod test_utils;
```
It was `#[doc(hidden)]` and documented that "External users are not supposed to rely on this module." So I really don't see the problem.

not-my-profile commented on src/parser.rs:
```diff
@@ -104,45 +105,49 @@ impl fmt::Display for ParserError {
     /// The index of the first unprocessed token in `self.tokens`
     index: usize,
-    dialect: &'a dyn Dialect,
+    dialect: PhantomData<D>,
```
rustc requires that generic type parameters of structs are actually used or it fails with a compile error telling you to use `PhantomData`. Note that `PhantomData` consumes no space it's just used by the compiler for the static analysis, see [core::marker::PhantomData](https://doc.rust-lang.org/core/marker/struct.PhantomData.html) and [PhantomData in the Nomicon](https://doc.rust-lang.org/nomicon/phantom-data.html).

not-my-profile commented on src/lib.rs:
```diff
@@ -21,14 +21,12 @@
 //!            ORDER BY a DESC, b";
 //!
-//! let ast = Parser::parse_sql(&dialect, sql).unwrap();
+//! let ast = Parser::<GenericDialect>::parse_sql(sql).unwrap();
```
I'd prefer to improve the documentation in a followup PR.

not-my-profile commented:
Hi Andrew, thanks for the review! I fixed the CI failure (it was just caused by `std::marker::PhantomData` not being available with `no_std`, I changed it to `core::marker::PhantomData`) and answered your comments :)

alamb commented:
Thanks again for this PR. I am closing it for now to clean up the PR list. Please reopen if that was a mistake.
